torch/_prims/utils.py

```diff
@@ -191,7 +205,89 @@ def compare_significant_strides(a: TensorLikeType, b: TensorLikeType):
         assert a.shape[idx] == b.shape[idx]
         if a.shape[idx] == 0 or a.shape[idx] == 1:
             continue
-        assert a.stride()[idx] == b.stride()[idx]
+        if a.stride()[idx] != b.stride()[idx]:
+            msg = (
+                "Stride mismatch! Strides are {0} and {1} (mismatched at {2})!".format(
+                    a.stride(), b.stride(), idx
+                )
+            )
+            raise RuntimeError(msg)
+
+
+# NOTE: Based on the implementation in TensorIterator.cpp, but note that
+# the note [Computing output strides] is incorrect, because it
+# says that strides will be preserved even if they are not
+# "non overlapping and dense", but this is incorrect. The
+# output of elementwise operations are always given
+# non overlapping and dense strides.
+def compute_elementwise_output_strides(*tensors) -> Tuple[int, ...]:
+    """
+    Computes the output strides for elementwise operations.
+    """
+
+    if len(tensors) == 0:
+        msg = "Can't compute elementwise output strides for zero tensors!"
+        raise ValueError(msg)
+
+    # Acquires the shape
+    check_same_shape(*tensors, allow_cpu_scalar_tensors=True)
+    tensors = tuple(
+        a for a in tensors if isinstance(a, TensorLike) and not is_cpu_scalar_tensor(a)
+    )
+
+    # Short-circuits for CPU scalar case
+    if len(tensors) == 0:
+        return ()
+
+    # Short-circuits for shapes with zero or one dimensions
+    # TODO: are these necessary?
+    ndim = tensors[0].ndim
+    if ndim == 0:
+        return ()
+    if ndim == 1:
+        return (1,)
+
+    shape = tensors[0].shape
+
+    def _cmp(idx_a, idx_b):
+        for tensor in tensors:
+            stride_a = tensor.stride()[idx_a]
+            stride_b = tensor.stride()[idx_b]
+
+            if stride_a == 0 or stride_b == 0:
+                continue
+
+            if stride_a < stride_b:
+                return -1
+
+            if stride_a > stride_b:
+                return 1
+
+            # stride_a == stride_b
+            if shape[idx_a] > shape[idx_b]:
+                return 1
+
+            # NOTE: this case is missing in the C++ impl
+            if shape[idx_a] < shape[idx_b]:
+                return -1
+
+        # Note: this case is hit if all strides are zero,
+        # or all strides are equal and all dimensions have the same length
+        return 0
+
+    perm = tuple(range(ndim))
+    perm = tuple(sorted(perm, key=cmp_to_key(_cmp), reverse=True))
+
+    permuted_shape = [-1] * ndim
+    for idx, x in enumerate(perm):
+        permuted_shape[idx] = shape[x]
+
+    new_strides = make_contiguous_strides_for(permuted_shape)
```

Comment:
nit: I'd expect permuted_strides correspond to permuted_shape, and what you are returning are output strides.